Library/ViewModels/PledgeViewModelTests.swift

```diff
@@ -1071,4 +1092,173 @@ final class PledgeViewModelTests: TestCase {
+      |> Project.lens.state .~ .live
+      |> Project.lens.personalization.isBacking .~ true
+      |> Project.lens.personalization.backing .~ (
+        .template
+          |> Backing.lens.paymentSource .~ GraphUserCreditCard.amex
+          |> Backing.lens.status .~ .errored
+          |> Backing.lens.reward .~ reward
+          |> Backing.lens.rewardId .~ reward.id
+          |> Backing.lens.shippingAmount .~ 10
+          |> Backing.lens.amount .~ 700
+      )
+
+    let mockService = MockService(
+      updateBackingResult: .failure(.invalidInput)
+    )
+
+    withEnvironment(apiService: mockService, currentUser: .template) {
+      self.vm.inputs.configureWith(project: project, reward: reward, refTag: nil, context: .update)
+      self.vm.inputs.viewDidLoad()
+
+      self.notifyDelegateUpdatePledgeDidSucceed.assertDidNotEmitValue()
+      self.updatePledgeFailedWithError.assertDidNotEmitValue()
+      self.confirmButtonEnabled.assertDidNotEmitValue()
+
+      self.vm.inputs.pledgeAmountViewControllerDidUpdate(
+        with: (amount: 10.0, min: 25.0, max: 10_000.0, isValid: true)
+      )
+
+      self.notifyDelegateUpdatePledgeDidSucceed.assertDidNotEmitValue()
+      self.updatePledgeFailedWithError.assertDidNotEmitValue()
+      self.confirmButtonEnabled.assertValues([true])
+
+      self.vm.inputs.confirmButtonTapped()
+
+      self.notifyDelegateUpdatePledgeDidSucceed.assertDidNotEmitValue()
+      self.updatePledgeFailedWithError.assertDidNotEmitValue()
+      self.confirmButtonEnabled.assertValues([true, false])
+
+      self.scheduler.run()
+
+      self.notifyDelegateUpdatePledgeDidSucceed.assertDidNotEmitValue()
+      self.updatePledgeFailedWithError.assertValueCount(1)
+      self.confirmButtonEnabled.assertValues([true, false, true])
+    }
+  }
+
+  func testConfirmButtonEnabled() {
+    let reward = Reward.postcards
+
+    let project = Project.cosmicSurgery
+      |> Project.lens.state .~ .live
+      |> Project.lens.personalization.isBacking .~ true
+      |> Project.lens.personalization.backing .~ (
+        .template
+          |> Backing.lens.paymentSource .~ GraphUserCreditCard.amex
+          |> Backing.lens.status .~ .errored
+          |> Backing.lens.reward .~ reward
+          |> Backing.lens.rewardId .~ reward.id
+          |> Backing.lens.shippingAmount .~ 10
+          |> Backing.lens.amount .~ 700
+      )
+
+    self.confirmButtonHidden.assertDidNotEmitValue()
+    self.confirmButtonEnabled.assertDidNotEmitValue()
+
+    self.vm.inputs.configureWith(project: project, reward: reward, refTag: nil, context: .update)
+    self.vm.inputs.viewDidLoad()
+
+    self.confirmButtonHidden.assertValues([false])
+    self.confirmButtonEnabled.assertDidNotEmitValue()
+
+    self.vm.inputs.pledgeAmountViewControllerDidUpdate(
+      with: (amount: 690, min: 25.0, max: 10_000.0, isValid: true)
+    )
+
+    self.confirmButtonHidden.assertValues([false])
+    self.confirmButtonEnabled.assertValues([false])
```

Comment:
Is it possible to add an assertion message here? Took me a while to figure out why the confirm button isn't enabled in this case 😅 (it's because the `amount` is actually just the same as was already pledged, right?).

Reply:
Sure will do :)